### PR TITLE
Add grpc server tracing

### PR DESCRIPTION
**Pull request by fedj**

### CODE REVIEW

File: ddtrace/contrib/grpc/server_interceptor.py [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+                raise
+
+    def extract_and_set_context(self, metadata):
+
+        context = self._propagator.extract(dict(metadata))
+        if context.trace_id:
+            self._pin.tracer.context_provider.activate(context)
+
+    def start_span(self, method):
+        return self._pin.tracer.trace('grpc.request', service=self._pin.service, resource=method, span_type='grpc')
```
Comment (brettlangdon) [inline ddtrace/contrib/grpc/server_interceptor.py]: missing newline at the end of the file

(most editors have a setting to do this for you automatically)

File: tests/contrib/grpc/test_grpc.py [hunk 2]
```diff
@@ -149,6 +157,34 @@ def test_pin_can_be_defined_per_channel(self):
         channel1.close()
         channel2.close()
 
+class GrpcServerTestCase(GrpcServerBaseMixin, unittest.TestCase):
```
Comment (brettlangdon) [inline tests/contrib/grpc/test_grpc.py]: We also should have tests that the appropriate `grpc.request` spans are being created for each request.

File: tests/contrib/grpc/test_grpc.py [hunk 3]
```diff
@@ -149,6 +157,34 @@ def test_pin_can_be_defined_per_channel(self):
...
+
+    def test_server_span_has_correct_parent_id(self):
```
Comment (brettlangdon) [inline tests/contrib/grpc/test_grpc.py]: A doctring here describing what this test is doing would be helpful, it took me a minute to figure out why we even had this test case.

Same goes for most other test cases too, we should probably have at least a quick one-liner description.

File: ddtrace/contrib/grpc/server_interceptor.py [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+        self._propagator = HTTPPropagator()
+
+    def intercept_service(self, continuation, handler_call_details):
+        if not self._pin or not self._pin.enabled():
+            return continuation(handler_call_details)
+
+        self.extract_and_set_context(handler_call_details.invocation_metadata)
+
+        with self.start_span(handler_call_details.method) as span:
+            try:
```
Comment (brettlangdon) [inline ddtrace/contrib/grpc/server_interceptor.py]: This `try/except` shouldn't be necessary, the span context manager will pick up on any exceptions and set the traceback appropriately

File: ddtrace/contrib/grpc/server_interceptor.py [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+                return continuation(handler_call_details)
+            except:
+                span.set_traceback()
+                raise
+
+    def extract_and_set_context(self, metadata):
+
+        context = self._propagator.extract(dict(metadata))
```
Comment (brettlangdon) [inline ddtrace/contrib/grpc/server_interceptor.py]: We probably want to have a configuration setting asking them if they want to enable distributed tracing or not (this is what we do in most other integrations right now).

Comment (brettlangdon) [reply ddtrace/contrib/grpc/server_interceptor.py]: Same likely goes for the client interceptor as well?

File: tests/contrib/grpc/test_grpc.py [hunk 6]
```diff
@@ -149,6 +157,34 @@ def test_pin_can_be_defined_per_channel(self):
...
+        with grpc.insecure_channel('localhost:%d' % (GRPC_PORT)) as channel:
+            stub = HelloStub(channel)
+            stub.SayHello(HelloRequest(name='test'))
+
+        server.stop(5)
+
+        writer = self._tracer.writer
+        spans = writer.pop()
+        eq_(len(spans), 2)
+        if (spans[0].name == 'grpc.client'):
```
Comment (brettlangdon) [inline tests/contrib/grpc/test_grpc.py]:
```suggestion
        if spans[0].name == 'grpc.client':
```

File: ddtrace/contrib/grpc/server_interceptor.py (hunk 5 above)
Comment (fedj) [reply ddtrace/contrib/grpc/server_interceptor.py]: Do you have any example of such configuration system ?

Comment (brettlangdon) [reply ddtrace/contrib/grpc/server_interceptor.py]: https://github.com/DataDog/dd-trace-py/blob/7db12f1c398c07cd5baf91c571aed672dbb6496d/ddtrace/contrib/requests/patch.py#L15-L20

https://github.com/DataDog/dd-trace-py/blob/7db12f1c398c07cd5baf91c571aed672dbb6496d/ddtrace/contrib/requests/connection.py#L63-L67

**Comment by brettlangdon** [comment]
@fedj any updates here? This PR has been open for a bit.

**Comment by fedj** [comment]
@brettlangdon Sorry I didn't have time to finish it. Can you wait on this ? If you can't wait, don't hesitate to take ownership of this PR

**Comment by fedj** [comment]
@brettlangdon Sorry I didn't have time to finish it. Can you wait on this ? If you can't wait, don't hesitate to take ownership of this PR

**Comment by fedj** [comment]
@brettlangdon Sorry I didn't have time to finish it. Can you wait on this ? If you can't wait, don't hesitate to take ownership of this PR

**Comment by brettlangdon** [comment]
@fedj not a problem at all! I figured that was the case, but didn't want to keep pushing it back if we didn't need to.

I will let you know if we take it over to finish.

**Comment by majorgreys** [comment]
@fedj We recently started PR #960 to add support for gRPC server following our other integrations. Please give that branch a try and tell us if it meets your needs for gRPC integration.

**Pull request closed without merging**